File: backend/src/main/java/com/ndoruhirwe/smartlogistics/service/PurchaseOrderService.java
package com.ndoruhirwe.smartlogistics.service;

import com.ndoruhirwe.smartlogistics.dto.request.PurchaseOrderCreateRequest;
import com.ndoruhirwe.smartlogistics.dto.request.PurchaseOrderItemCreateRequest;
import com.ndoruhirwe.smartlogistics.dto.request.PurchaseOrderItemUpdateRequest;
import com.ndoruhirwe.smartlogistics.dto.request.PurchaseOrderUpdateRequest;
import com.ndoruhirwe.smartlogistics.dto.response.PurchaseOrderResponse;
import com.ndoruhirwe.smartlogistics.entity.enums.PurchaseOrderStatus;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.UUID;

public interface PurchaseOrderService {

    PurchaseOrderResponse create(PurchaseOrderCreateRequest request);

    PurchaseOrderResponse getById(UUID id);

    Page<PurchaseOrderResponse> getAll(Pageable pageable);

    Page<PurchaseOrderResponse> getByStatus(PurchaseOrderStatus status,
            Pageable pageable);

    Page<PurchaseOrderResponse> getBySupplier(UUID supplierId, Pageable pageable);

    Page<PurchaseOrderResponse> getByWarehouse(UUID warehouseId, Pageable pageable);

    PurchaseOrderResponse update(UUID id, PurchaseOrderUpdateRequest request);

    PurchaseOrderResponse addItem(UUID purchaseOrderId,
            PurchaseOrderItemCreateRequest request);

    PurchaseOrderResponse updateItem(UUID purchaseOrderId, UUID itemId,
            PurchaseOrderItemUpdateRequest request);

    PurchaseOrderResponse removeItem(UUID purchaseOrderId, UUID itemId);

    PurchaseOrderResponse submit(UUID id);

    PurchaseOrderResponse approve(UUID id);

    PurchaseOrderResponse cancel(UUID id);
}

File: frontend/src/app/features/procurement/services/purchase-order.service.ts
import { inject, Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';

import { API_ENDPOINTS } from '../../../core/constants/api-endpoints';
import { PageResponse } from '../../../core/models/page-response.model';

import { PurchaseOrderResponse } from '../models/purchase-order-response';
import { PurchaseOrderCreateRequest } from '../models/purchase-order-create-request';
import { PurchaseOrderUpdateRequest } from '../models/purchase-order-update-request';
import { PurchaseOrderItemCreateRequest } from '../models/purchase-order-item-create-request';
import { PurchaseOrderItemUpdateRequest } from '../models/purchase-order-item-update-request';
import { PurchaseOrderStatus } from '../models/purchase-order-status';

@Injectable({
  providedIn: 'root'
})
export class PurchaseOrderService {

  private readonly http = inject(HttpClient);

  getAll(
    page = 0,
    size = 10
  ): Observable<PageResponse<PurchaseOrderResponse>> {

    const params = new HttpParams()
      .set('page', page)
      .set('size', size);

    return this.http.get<PageResponse<PurchaseOrderResponse>>(
      API_ENDPOINTS.purchaseOrders,
      { params }
    );
  }

  getById(
    id: string
  ): Observable<PurchaseOrderResponse> {

    return this.http.get<PurchaseOrderResponse>(
      `${API_ENDPOINTS.purchaseOrders}/${id}`
    );
  }

  create(
    request: PurchaseOrderCreateRequest
  ): Observable<PurchaseOrderResponse> {

    return this.http.post<PurchaseOrderResponse>(
      API_ENDPOINTS.purchaseOrders,
      request
    );
  }

  update(
    id: string,
    request: PurchaseOrderUpdateRequest
  ): Observable<PurchaseOrderResponse> {

    return this.http.put<PurchaseOrderResponse>(
      `${API_ENDPOINTS.purchaseOrders}/${id}`,
      request
    );
  }

  addItem(
    purchaseOrderId: string,
    request: PurchaseOrderItemCreateRequest
  ): Observable<PurchaseOrderResponse> {

    return this.http.post<PurchaseOrderResponse>(
      `${API_ENDPOINTS.purchaseOrders}/${purchaseOrderId}/items`,
      request
    );
  }

  updateItem(
    purchaseOrderId: string,
    itemId: string,
    request: PurchaseOrderItemUpdateRequest
  ): Observable<PurchaseOrderResponse> {

    return this.http.put<PurchaseOrderResponse>(
      `${API_ENDPOINTS.purchaseOrders}/${purchaseOrderId}/items/${itemId}`,
      request
    );
  }

  removeItem(
    purchaseOrderId: string,
    itemId: string
  ): Observable<PurchaseOrderResponse> {

    return this.http.delete<PurchaseOrderResponse>(
      `${API_ENDPOINTS.purchaseOrders}/${purchaseOrderId}/items/${itemId}`
    );
  }

  submit(
    id: string
  ): Observable<PurchaseOrderResponse> {

    return this.http.patch<PurchaseOrderResponse>(
      `${API_ENDPOINTS.purchaseOrders}/${id}/submit`,
      {}
    );
  }

  approve(
    id: string
  ): Observable<PurchaseOrderResponse> {

    return this.http.patch<PurchaseOrderResponse>(
      `${API_ENDPOINTS.purchaseOrders}/${id}/approve`,
      {}
    );
  }

  cancel(
    id: string
  ): Observable<PurchaseOrderResponse> {

    return this.http.patch<PurchaseOrderResponse>(
      `${API_ENDPOINTS.purchaseOrders}/${id}/cancel`,
      {}
    );
  }

  getByStatus(
    status: PurchaseOrderStatus,
    page = 0,
    size = 10
  ): Observable<PageResponse<PurchaseOrderResponse>> {

    const params = new HttpParams()
      .set('status', status)
      .set('page', page)
      .set('size', size);

    return this.http.get<PageResponse<PurchaseOrderResponse>>(
      `${API_ENDPOINTS.purchaseOrders}/status`,
      { params }
    );
  }
}
